middleware/src/middleware.go

```diff
@@ -335,56 +381,211 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
+// The function first validates the current password with redis, then replaces it with the new password.
+// Passwords need to be longer than or equal to 8 chars.
 func (middleware *Middleware) UserChangePassword(args rpcmessages.UserChangePasswordArgs) rpcmessages.ErrorResponse {
-	if len(args.NewPassword) >= 8 {
-		if args.Username == "admin" {
-			middleware.dummyAdminPassword = args.NewPassword
-			if !middleware.dummyIsBaseSetup {
-				middleware.dummyIsBaseSetup = true // the change of the admin password completes the dummy setup process (for now)
-			}
+	if len(args.NewPassword) < 8 {
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "password change unsuccessful, the password needs to be at least 8 characters in length",
+			Code:    rpcmessages.ErrorPasswordTooShort,
 		}
-		return rpcmessages.ErrorResponse{Success: true}
 	}
 
-	return rpcmessages.ErrorResponse{
-		Success: false,
-		Message: "password change unsuccessful (too short)",
-		Code:    rpcmessages.ErrorDummyPasswordTooShort,
+	usersMap := make(map[string]UserAuthStruct)
+	authStructureString, err := middleware.redisClient.GetString(redis.MiddlewareAuth)
+	if err != nil {
+		log.Println("error when getting the authentication structure from redis")
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "password change unsuccessful",
+			Code:    rpcmessages.ErrorPasswordChangeFailed,
+		}
+	}
+	err = json.Unmarshal([]byte(authStructureString), &usersMap)
+	if err != nil {
+		log.Println("Did not receive json from redis's authentication structure")
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "password change unsuccessful",
+			Code:    rpcmessages.ErrorPasswordChangeFailed,
+		}
+	}
+
+	//TODO: Once we support multiple users work over the ErrorPasswordChangeUsernameNotExist message. It reveals information about the database.
+	if _, ok := usersMap[args.Username]; !ok {
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "username does not exist",
+			Code:    rpcmessages.ErrorPasswordChangeUsernameNotExist,
+		}
+	}
+
+	passwordFromStorage := usersMap[args.Username]
+	err = bcrypt.CompareHashAndPassword([]byte(passwordFromStorage.BCryptedPassword), []byte(args.Password))
+	if err != nil {
+		log.Println("Hash and password did not match")
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "password change unsuccessful, current password was incorrect",
+			Code:    rpcmessages.ErrorPasswordChangePasswordIncorrect,
+		}
+	}
+
+	bcryptedPassword, err := bcrypt.GenerateFromPassword([]byte(args.NewPassword), 12)
+	if err != nil {
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "password change unsuccessful",
+			Code:    rpcmessages.ErrorPasswordChangeFailed,
+		}
 	}
+
+	userAuthSecrets := usersMap[args.Username]
+	userAuthSecrets.BCryptedPassword = string(bcryptedPassword)
+	userAuthSecrets.Role = "admin"
```

Comment:
Is the role secret? :D

Reply:
Depends on the ingredients, I like cinnamon in my role.